pkg/planner/core/operator/logicalop/logical_datasource.go

```diff
@@ -603,72 +606,112 @@ func preferKeyColumnFromTable(dataSource *DataSource, originColumns []*expressio
 				continue
 			}
-			onlyLogicOpAndFTS := expression.ExprOnlyContainsLogicOpAndFTS(cond)
-			if containsFTS && !onlyLogicOpAndFTS {
-				return plannererrors.ErrWrongUsage.FastGen(plannererrors.FTSWrongPlace)
+		}
+		if matchedIndexIsHinted && !path.Forced {
+			// If we have found a hinted index matched before, we should skip
+			// the non-hinted indexes later.
+			continue
+		}
+		ftsCols := intset.NewFastIntSet()
+		invertedIndexedCols := intset.NewFastIntSet()
+		if path.Index.FullTextInfo != nil {
+			for _, indexCol := range path.Index.Columns {
+				col := ds.TableInfo.Columns[indexCol.Offset]
+				ftsCols.Insert(int(col.ID))
+			}
+		}
+		if path.Index.HybridInfo != nil {
+			for _, ftsInfo := range path.Index.HybridInfo.FullText {
+				for _, indexCol := range ftsInfo.Columns {
+					col := ds.TableInfo.Columns[indexCol.Offset]
+					ftsCols.Insert(int(col.ID))
+				}
+			}
+			for _, invertedInfo := range path.Index.HybridInfo.Inverted {
+				for _, indexCol := range invertedInfo.Columns {
+					col := ds.TableInfo.Columns[indexCol.Offset]
+					invertedIndexedCols.Insert(int(col.ID))
+				}
 			}
 		}
-		idSetForCheck.Clear()
-		expression.CollectColumnIDForFTS(sf, &idSetForCheck)
-		// Check TiCI version first.
-		var currentIndex *model.IndexInfo
-		for idx, set := range ticiIdx2FastCheck {
-			// The used columns in the FTS function should be a subset of the index columns.
-			if idSetForCheck.SubsetOf(set) {
-				currentIndex = idx
-				goto finalCheck
+		noUnmatchedFTSFunc := true
+		tmpMatchedExprSet.Clear()
+	checkExprForIndexLoop:
+		for i, cond := range ds.PushedDownConds {
+			fullyCovered := expression.ExprCoveredByOneTiCIIndex(cond, &ftsCols, &invertedIndexedCols)
+			if !fullyCovered {
+				if expression.ContainsFullTextSearchFn(cond) {
+					noUnmatchedFTSFunc = false
+					break checkExprForIndexLoop
+				}
+				continue
+			}
+			if fullyCovered {
+				tmpMatchedExprSet.Insert(i)
 			}
 		}
-	finalCheck:
-		// If the index is not found, it means that the FTS function is not valid.
-		if currentIndex == nil {
-			return errors.New("Full text search can only be used with a matching fulltext index and a columnar storage")
+		if !noUnmatchedFTSFunc {
+			continue
 		}
-		// Currently TiDB doesn't support multiple fulltext search functions used with multiple index calls.
-		if matchedIdx != nil && matchedIdx.ID != currentIndex.ID {
-			return errors.New("Current TiDB doesn't support multiple fulltext search functions used with multiple index calls")
+		hasUnmatchedFTSOverAllIdx = false
+		// We have filterer out (!path.Forced && matchedIndexIsHinted) case before.
+		// So after we check the (path.Forced && !matchedIndexIsHinted) case, the implicit case is:
+		// !path.Forced && !matchedIndexIsHinted or path.Forced && matchedIndexIsHinted
+		if (path.Forced && !matchedIndexIsHinted) || tmpMatchedExprSet.Len() > matchedExprSetForChosenIndex.Len() {
```

Comment:
why every time we need tmpMatchedExprSet.Len() > matchedExprSetForChosenIndex.Len()？

Reply:
This is an easy rule-based index selection.